Support for background gradients and images in CupertinoSliverNavigationBar and CupertinoNavigationBar

benjaminhorner opened the pull request:
## Description
Adding support for **gradient** backgrounds and **image** backgrounds in `CupertinoSliverNavigationBar` and `CupertinoNavigationBar`.

Background gradients may be added like so:

```
CupertinoSliverNavigationBar(
     backgroundGradient: LinearGradient(colors: [
         Color(0xf44141).withOpacity(1),
          Color(0x9d41f4).withOpacity(1),
      ]),
      largeTitle: Text("Large title item 1"),
),
```

Here is the result:
![Simulator Screen Shot - iPhone Xs - 2019-05-22 at 11 44 38](https://user-images.githubusercontent.com/3463436/58164945-136f4800-7c87-11e9-9351-a18e8244f694.png)
![Simulator Screen Shot - iPhone Xs - 2019-05-22 at 11 44 36](https://user-images.githubusercontent.com/3463436/58164999-271aae80-7c87-11e9-9179-343405c5925c.png)

Background images may be added like so:

```
CupertinoSliverNavigationBar(
      backgroundImage: NetworkImage("https://api.adorable.io/avatars/285/abott@adorable.png"),
      largeTitle: Text("Large title item 1"),
),
```

Here is the result:
![Simulator Screen Shot - iPhone Xs - 2019-05-22 at 11 44 25](https://user-images.githubusercontent.com/3463436/58165029-3863bb00-7c87-11e9-92d8-a2d5700a577c.png)
![Simulator Screen Shot - iPhone Xs - 2019-05-22 at 11 44 28](https://user-images.githubusercontent.com/3463436/58165030-3863bb00-7c87-11e9-8fae-48e8beba33da.png)

## Related Issues

https://github.com/flutter/flutter/issues/25144

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

benjaminhorner committed:
Support for background gradients and images in CupertinoSliverNavigationBar and CupertinoNavigationBar

benjaminhorner committed:
Take into account null image provider

xster commented:
Thanks for your contribution. The general concept is sensible. Though rather than splitting out all the subcomponents of the BoxDecoration, see the Container implementation. We can let the user override a `decoration` instead. The color parameter can still be a shorthand for just overriding the color and if you supply both, color gets ignored.

benjaminhorner commented:
I will dig into that idea.
I also will look into adding animation tween for gradients that would differ from one navigation bar to the next.
Brian Egan has a good article about this: https://flutterwidgets.io/animating-gradients.html

benjaminhorner committed:
Added LinearGradientTween

benjaminhorner committed:
Added support for BoxDecortion ’background’ with LinearGradientTween

benjaminhorner commented:
@xster I took the direction of your suggestion with having a `background` property on the `CupertinoNavigationBar` and `CupertinoSliverNavigationBar` that takes a `BoxDecoration`.

I also added support for `LinearGradientTween` which smoothly animates from one gradient to the other between views.

I am unsure though on how to handle the case where one `CupertinoSliverNavigationBar` would have a `background` with a **gradient** and another with a `backgroundColor`. As it stands the animation **wouldn't be smooth**.
My thought in cases where the background properties to animate are different, fallback to a cross fade or something. But I am unsure how to do that.

What do you think?

benjaminhorner commented:
@xster I came up with a solution for animating between navigation bars with **gradients** and navigation bars with a `backgroundColor`:
if one of the navigation bars does not have a gradient set, I create a `LinearGradient` with the `backgroundColor` and animate between that and the gradient background. It feels a little hacky to me but I am not sure this can be done any other way.

I think a similar result could be achieved for **background images** where one would fade the image in/out hiding/revealing the `backgroundColor`. I am working on this.

And when animating between to images, just a simple cross fade.

benjaminhorner committed:
Added support for transition between gradients and solid color backgrounds

benjaminhorner commented:
TODO: 
- [ ] animate between two background images
- [ ] animate between a background image and a gradient
- [ ] animate between a background image and a color

benjaminhorner committed:
Cleanup + support for border tween

benjaminhorner committed:
Fox for background wsith color set but no gradient

benjaminhorner committed:
Added all background image transitions

benjaminhorner committed:
Don’t force BoxFit.cover for background images

benjaminhorner commented:
@xster: I am done adding the support for background images also.
I have added all transitions too. 

Transitions support:
color -> color
color -> gradient
color -> image
gradient -> gradient
gradient -> image
image -> image

Here is an example implementation:

Color:
```
CupertinoSliverNavigationBar(
    backgroundColor: Colors.blue,
    largeTitle: Text("Large title item 1"),
),
```

OR Color:
```
CupertinoSliverNavigationBar(
    background: BoxDecoration(
        border: Border(bottom: BorderSide(color: Colors.blue)),
        color: Colors.red,
    ),
largeTitle: Text("Large title item 1"),
),
```
Gradient:
```
CupertinoSliverNavigationBar(
    background: BoxDecoration(
        gradient: LinearGradient(colors: [
            Colors.orange.withOpacity(1),
            Colors.green.withOpacity(1),
        ]),
    ),
largeTitle: Text("Large title item 1"),
),
```
Image:
```
CupertinoSliverNavigationBar(
    background: BoxDecoration(
        image: DecorationImage(
            image: NetworkImage("https://api.adorable.io/avatars/285/abott@adorable.png"),
            fit: BoxFit.cover, // any BoxFit type works here
        ),
     ),
largeTitle: Text("Large title item 1"),
),
```

I see some tests are failing. I will have to check that but I would love to have your feedback and/or suggestions on how to improve it.

I have also noticed that during transitions, the color of the chevron and actions icons stay blue whatever the theme. But that will go into another PR as this is not related to my current code/PR.

benjaminhorner committed:
Compute luminance for gradients

benjaminhorner committed:
Added support for background color luminance

benjaminhorner commented:
I added support for computed luminance of gradient backgrounds. 
My idea was to calculate the average luminance of the colors contained in the gradient's colors array. I decided to give more importance to the first and last colors as they are closest to the icons in the status bar (I doubled their value).

I also added the ability for the developer to override the computed luminance by adding a `color` property to the `background` whose computed luminance value takes precedence over the gradient's luminance value.

gspencergoog commented:
@HansMuller or @xster, can one of you please take another look at this PR? It's been languishing for a while now...

@benjaminhorner, can you please address the code conflicts and the test failures?  Those will need to be addressed before we can accept the PR in any case.

benjaminhorner committed:
Merge branch 'master' into master

benjaminhorner committed:
Merge branch 'master' of https://github.com/flutter/flutter

# Conflicts:
#	packages/flutter/lib/src/cupertino/nav_bar.dart

benjaminhorner committed:
Merge branch 'master' of https://github.com/benjaminhorner/flutter

# Conflicts:
#	packages/flutter/lib/src/cupertino/nav_bar.dart

benjaminhorner commented:
@gspencergoog I see all tests are failing `Because platform_views depends on path_provider >=1.4.5 which requires Flutter SDK version >=1.9.1+hotfix.5 <2.0.0, version solving failed.` but I must say I am unsure how to fix this. It doesn't seem related to my PR nevertheless the tests fail. Any suggestion on where or what to look for?

gspencergoog commented on packages/flutter/lib/src/animation/tween.dart:
```diff
@@ -5,6 +5,7 @@
 import 'dart:ui' show Color, Size, Rect;
 
 import 'package:flutter/foundation.dart';
+import 'package:flutter/src/painting/gradient.dart';
```
The normal way we do things is to import the sub-package, not the individual file:
```dart
import 'package:flutter/painting.dart';
```

gspencergoog commented:
I agree that the version solving failure doesn't seem related to your PR. Perhaps you just synced to a broken revision on flutter/flutter master?  Maybe try syncing again. When I try your change locally at HEAD, it doesn't have the version solving issues (although it does have other analyzer failures).

You can run the analyzer locally with `flutter analyze --flutter-repo`.

Also, once that's resolved, you're going to need to add some tests for the code changes you made, since that's part of our [tree hygiene requirements](https://github.com/flutter/flutter/wiki/Tree-hygiene).

And I'd suggest reverting all of the whitespace-only changes you made: likely that is dartfmt formatting things: the Flutter tree doesn't use dartfmt. It also makes it harder to see what the substantive changes are.

benjaminhorner committed:
Merge branch 'master' of https://github.com/flutter/flutter

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -87,13 +88,46 @@ class _HeroTag {
+        return background.color.computeLuminance() < luminanceThreshold;
+      } else if (background?.gradient != null) {
+        double gradientLuminanceThreshold = 0.0;
+        gradientLuminanceThreshold +=
```
why do you need to special case the first and last ones? Just the for loop should be fine no?

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -112,14 +146,14 @@ Widget _wrapWithBackground({
   }
   final DecoratedBox childWithBackground = DecoratedBox(
-    decoration: BoxDecoration(
+    decoration: background != null ? background : BoxDecoration(
```
This can just be `background ?? BoxDecoration(...`

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -215,25 +250,37 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
     this.trailing,
     this.border = _kDefaultNavBarBorder,
     this.backgroundColor,
+    this.background,
```
For consistency (with classes like `Container`) I'd call this decoration rather than background.

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -215,25 +250,37 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
-       ),
-       super(key: key);
+  })
+      : assert(automaticallyImplyLeading != null),
```
I'd avoid extraneous re-formatting with this PR.

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -215,25 +250,37 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
+        assert(automaticallyImplyMiddle != null),
+        assert(transitionBetweenRoutes != null),
+        assert(
+        heroTag != null,
```
Keep the existing indentation patterns

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -311,6 +358,13 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
   final Color backgroundColor;
 
+  /// {@template flutter.cupertino.navBar.background}
+  /// The background of the navigation bar.
```
`A [BoxDecoration] drawn behind the navigation bar content.`

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -112,14 +146,14 @@ Widget _wrapWithBackground({
+  if (backgroundColor?.alpha == 0xFF)
     return childWithBackground;
 
   return ClipRect(
```
You'd have to handle this case too. If you create a custom box decoration with a translucent background color, it should be blurred.

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -311,6 +358,13 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
 
+  /// {@template flutter.cupertino.navBar.background}
+  /// The background of the navigation bar.
+  /// If set, the [backgroundColor] and [border] properties will be ignored.
```
Create reverse doc references too in backgroundColor and border.

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -311,6 +358,13 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
+  /// {@template flutter.cupertino.navBar.background}
+  /// The background of the navigation bar.
+  /// If set, the [backgroundColor] and [border] properties will be ignored.
+  ///
```
remove extra line

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -485,9 +543,16 @@ class _CupertinoNavigationBarState extends State<CupertinoNavigationBar> {
-              titleTextStyle: CupertinoTheme.of(context).textTheme.navTitleTextStyle,
+              backgroundColor: widget.backgroundColor,
+              background: widget.background,
+              backButtonTextStyle: CupertinoTheme
```
avoid extraneous reformatting for untouched lines.

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -673,13 +753,15 @@ class CupertinoSliverNavigationBar extends StatefulWidget {
 
   @override
-  _CupertinoSliverNavigationBarState createState() => _CupertinoSliverNavigationBarState();
+  _CupertinoSliverNavigationBarState createState() =>
```
ditto

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -737,12 +820,14 @@ class _CupertinoSliverNavigationBarState extends State<CupertinoSliverNavigation
 
 class _LargeTitleNavigationBarSliverDelegate
-    extends SliverPersistentHeaderDelegate with DiagnosticableTreeMixin {
+    extends SliverPersistentHeaderDelegate
```
avoid mass reformatting

xster commented on packages/flutter/lib/src/cupertino/nav_bar.dart:
```diff
@@ -1559,25 +1668,50 @@ class _NavigationBarTransition extends StatelessWidget {
+    begin: bottomNavBar.renderBox.size.height,
+    end: topNavBar.renderBox.size.height,
+  ),
+        backgroundTween = LinearGradientTween(
```
rather than cherry-picking contents of the box decoration, you should just tween between the box decoration instance altogether. You can see that BoxDecoration.lerp already lerps Gradient.lerp. Generic logic is preferred.

xster reviewed (commented):
Thanks for your contribution! Sorry we dropped the ball on this review. 

Some general feedback:

- You might have to rebase the branch on flutter/flutter's master again to avoid the lint issue
- Avoid a global reformat. It'll help make the review easier if affected lines are related to the issue
- The PR needs tests for new additions. See existing tests for patterns

himanshusharma89 commented:
Hey, I tried the LinearGradientTween from this PR and it worked great for me, thank you @benjaminhorner
And it would be great if we have this tween officially.

dnfield commented:
Since there has not been a response on this PR in some time, and it's not even mergable in its current state, I'm going to close it. Please feel free to ping the PR again if you plan to continue working on it.
